Refactor core learner

canerturkmen opened the pull request:
*Issue #, if available:*

*Description of changes:*

Factors out an `AbstractLearner` class to `autogluon.core`, mainly consisting serialization/deserialization methods, and adding stubs for `fit`, `predict`, `score`, `leaderboard`. This is mainly to prevent code dupe in forecasting.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

szha commented:
Job PR-1733-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1733/1/index.html

canerturkmen commented on core/src/autogluon/core/learner/abstract_learner.py:
```diff
@@ -0,0 +1,127 @@
+        save_path = path_context + self.learner_file_name
+        return path_context, model_context, save_path
+
+    def set_contexts(self, path_context):
```
This carries over from the Tabular learner, where `path_context` is thought of as the top-level directory where all models, learner, etc. will be stored. Added docstrings for this and slightly improved methods.

Innixma reviewed (commented):
Looks good! Had a few comments. We will want to wait for v0.4.1 release early next week before merging as this touches core and tabular.

canerturkmen commented on tabular/src/autogluon/tabular/learner/abstract_learner.py:
```diff
@@ -802,36 +747,3 @@ def distill(self, X=None, y=None, X_val=None, y_val=None, time_limit=None, hyper
-        save_json.save(path=self.path + self.learner_info_json_name, obj=info)
-        return info
-
-    # TODO: Add data info gathering at beginning of .fit() that is used by all learners to add to get_info output
-    # TODO: Add feature inference / feature engineering info to get_info output
-    def get_info(self, **kwargs):
-        learner_info = {
-            'path': self.path,
-            'label': self.label,
-            'random_state': self.random_state,
-            'version': self.version,
-            'features': self.features,
-            'feature_metadata_in': self.feature_metadata_in,
-        }
```
Ah careless mistake. Thanks for catching this.

canerturkmen commented:
Rev 2 addresses comments from Nick. Thanks! I had initially intended the abstract constructor just to define the instance variables types, such that things would work even if subclasses did not call `super().__init__`, but an actual abstract constructor is probably better.

szha commented:
Job PR-1733-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1733/3/index.html

Innixma reviewed (approved):
LGTM! Thanks for the updates, this will be very valuable to avoid duplicated code.
